feat(curriculum): adding content for HTML and Accessibility review page (#57021)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/review-html-accessibility/671a87a39b88245a579c2271.md

## Assistive Technology for Accessibility

- **Screen readers**: software programs that read the content of a computer screen out loud. They are used by people who are blind or visually impaired to access the web.
- **Large text or braille keyboards**: used by people with visual impairments to access the web.
- **Screen magnifiers**: software programs that enlarge the content of a computer screen. They are used by people with low vision to access the web.
- **Alternative pointing devices**: used by people with mobility impairments to access the web. This includes devices such as joysticks, trackballs, and touchpads.
- **Voice recognition**: used by people with mobility impairments to access the web. It allows users to control a computer using their voice.

## Accessibility Auditing Tools

- **Common Accessibility Tools**: Google Lighthouse, Wave, IBM Equal Accessibility Checker, and axe DevTools are some of the common accessibility tools used to audit the accessibility of a website.

## Accessibility Best Practices

> - **Proper heading level structure**: You should use proper heading levels to create a logical structure for your content. This helps screen readers and other assistive technologies understand the content of your website.

Reviewer: ```suggestion
- **Proper heading level structure**: You should use proper heading levels to create a logical structure for your content. This helps people using screen readers and other assistive technologies understand the content of your website.
```

I apologize for being so picky here. I just want to add the emphasis that this helps people, not technology.

> - **Accessibility and Tables**: When using tables, you should use the `<th>` element to define header cells and the `<td>` element to define data cells. This helps screen readers and other assistive technologies understand the structure of the table.

Reviewer: ```suggestion
- **Accessibility and Tables**: When using tables, you should use the `<th>` element to define header cells and the `<td>` element to define data cells. This helps people using screen readers and other assistive technologies understand the structure of the table.
```

Again, sorry for the last minute suggestions.

> - **Importance for inputs to have an associated label**: You should use the `<label>` element to associate labels with form inputs. This helps screen readers and other assistive technologies understand the purpose of the input.

Reviewer: ```suggestion
- **Importance for inputs to have an associated label**: You should use the `<label>` element to associate labels with form inputs. This helps people using screen readers and other assistive technologies understand the purpose of the input.
```

> - **Importance of good `alt` text**: You should use the `alt` attribute to provide a text alternative for images. This helps screen readers and other assistive technologies understand the content of the image.

Reviewer: ```suggestion
- **Importance of good `alt` text**: You should use the `alt` attribute to provide a text alternative for images. This helps people using screen readers and other assistive technologies understand the content of the image.
```

> - **Importance of good link text**: You should use descriptive link text to help users understand the purpose of the link. This helps screen readers and other assistive technologies understand the content of the link.

Reviewer: ```suggestion
- **Importance of good link text**: You should use descriptive link text to help users understand the purpose of the link. This helps people using screen readers and other assistive technologies understand the content of the link.
```

- **Best practices for making audio and video content accessible**: You should provide captions and transcripts for audio and video content to make it accessible to people with hearing impairments. You should also provide audio descriptions for video content to make it accessible to people with visual impairments.

> - **`tabindex` attribute**: used to make elements focusable and define the relative order in which they should be navigated using the keyboard. It important to never use the `tabindex` attribute with a value greater than 0. Instead, you should either use a value of 0 or -1. For more information, review the prior lecture video on keyboard accessibility.

Reviewer: ```suggestion
- **`tabindex` attribute**: used to make elements focusable and define the relative order in which they should be navigated using the keyboard. It is important to never use the `tabindex` attribute with a value greater than 0. Instead, you should either use a value of 0 or -1. For more information, review the prior lecture video on keyboard accessibility.
```

- **`accesskey` attribute**: used to define a keyboard shortcut for an element. This can help users with mobility impairments navigate the website more easily.

## WAI-ARIA, Roles, and Attributes

- **WAI-ARIA**: a set of attributes that can be added to HTML elements to improve accessibility. It provides additional information to assistive technologies about the purpose and structure of the content.

> - **Aria roles**: a set of predefined roles that can be added to HTML elements to define their purpose. This helps screen readers and other assistive technologies understand the content of the website. Examples include `role="tab"`, `role="menu"`, and `role="alert"`.

Reviewer: ```suggestion
- **Aria roles**: a set of predefined roles that can be added to HTML elements to define their purpose. This helps people using screen readers and other assistive technologies understand the content of the website. Examples include `role="tab"`, `role="menu"`, and `role="alert"`.
```

- **`aria-label` and `aria-labelledby` attributes**: These attributes are used to give an element a programmatic (or accessible) name, which helps people using assistive technology (such as screen readers) understand the purpose of the element. They are often used when the visual label for an element is an image or symbol rather than text. `aria-label` allows you to define the name directly in the attribute while `aria-labelledby` allows you to reference existing text on the page.

> - **`aria-hidden` attribute**: Used to hide an element from assistive technology such as screen readers. For example, this can be used to hide decorative images that do not provide any meaningful content.

Reviewer: ```suggestion
- **`aria-hidden` attribute**: Used to hide an element from people using assistive technology such as screen readers. For example, this can be used to hide decorative images that do not provide any meaningful content.
```

- **`aria-expanded` attribute**: used to convey the state of a toggle (or disclosure) feature to screen reader users.
- **`aria-live` attribute**: used to indicate that an element's content is important enough to require that any changes to the content should be announced immediately to screen reader users. This can include status messages like updating the number of results returned from a search, or an error message displayed after an unsuccessful form submission.

> - **Common Aria states**: Common Aria states include `aria-haspopup`, `aria-checked`, `aria-disabled`, and `aria-selected`. These attributes can be used to indicate the state of an element and help screen readers and other assistive technologies understand the content of the website.

Reviewer: ```suggestion
- **Common Aria states**: Common Aria states include `aria-haspopup`, `aria-checked`, `aria-disabled`, and `aria-selected`. These attributes can be used to indicate the state of an element and help people using screen readers and other assistive technologies understand the content of the website.
```

> - **`aria-controls` attribute**: used to associate an element with another element that it controls. This helps screen readers and other assistive technologies understand the relationship between the elements.

Reviewer: ```suggestion
- **`aria-controls` attribute**: used to associate an element with another element that it controls. This helps people using screen readers and other assistive technologies understand the relationship between the elements.
```

> - **`aria-describedby` attribute**: used to associate an element with another element that describes it. This helps screen readers and other assistive technologies understand the purpose of the element.

Reviewer: ```suggestion
- **`aria-describedby` attribute**: used to associate an element with another element that describes it. This helps people using screen readers and other assistive technologies understand the purpose of the element.
```

Review the concepts below to prepare for the upcoming quiz.

## Introduction to Accessibility

- **Web Content Accessibility Guidelines**: The Web Content Accessibility Guidelines (WCAG) are a set of guidelines for making web content more accessible to people with disabilities. The four principles of WCAG are POUR which stands for Perceivable, Operable, Understandable, and Robust.

## Assistive Technology for Accessibility

- **Screen readers**: software programs that read the content of a computer screen out loud. They are used by people who are blind or visually impaired to access the web.
- **Large text or braille keyboards**: used by people with visual impairments to access the web.
- **Screen magnifiers**: software programs that enlarge the content of a computer screen. They are used by people with low vision to access the web.
- **Alternative pointing devices**: used by people with mobility impairments to access the web. This includes devices such as joysticks, trackballs, and touchpads.
- **Voice recognition**: used by people with mobility impairments to access the web. It allows users to control a computer using their voice.

> ## Accessibility Auditing Tools

Reviewer: Just a suggestion, I think you are leaving out by far the most popular automated accessibility testing tool, [axe DevTools](https://www.deque.com/axe/devtools/). In fact, that's what Lighthouse uses underneath the hood to do its testing. And it is also works in all browsers.

Author: I pulled this list of tools from the existing tools mention in the lectures. We could add the `axe DevTools` to the list but we will need to add a section for that in the corresponding lecture script.  It looks like the videos for this lecture block haven't been recorded yet, so I will need to alert the person recording this about this change

Reviewer: Ya, no worries. I figured this list was dependent upon was was being taught. I just wanted to mention it because axe is by far the most popular tool for accessibility testing and nobody I know in accessibility uses Lighthouse.

@jdwilkin4, a bit off topic, I apologize, but is it possible to review the lecture scripts for this?

Author: Absolutely!

If you want to message me your email through private message on the forum or discord, I can send you the actual scripts copied to a Google doc so you can leave notes/or edit the copies directly. Then we can update the actual scripts.

The person recording the accessibility videos has been notified to hold off on recording these set of videos

I will give you list for the HTML and Accessibility lecture block as well as the CSS and Accessibility lecture block

- **Importance for inputs to have an associated label**: You should use the `<label>` element to associate labels with form inputs. This helps screen readers and other assistive technologies understand the purpose of the input.
- **Importance of good `alt` text**: You should use the `alt` attribute to provide a text alternative for images. This helps screen readers and other assistive technologies understand the content of the image.
- **Importance of good link text**: You should use descriptive link text to help users understand the purpose of the link. This helps screen readers and other assistive technologies understand the content of the link.
- **Best practices for making audio and video content accessible**: You should provide captions and transcripts for audio and video content to make it accessible to people with hearing impairments. You should also provide audio descriptions for video content to make it accessible to people with visual impairments.

> - **`tabindex` attribute**: used to specify the order in which elements should receive focus when using the tab key to navigate the website. This can help users with mobility impairments navigate the website more easily.

Reviewer: I am concerned that the explanation for `tabindex` is giving the wrong impression and encouraging improper use. Basically, you should never use `tabindex` to specify the order that elements should receive focus (in other words, you should never use a `tabindex` value greater than `0`). The order that elements should receive focus should be determined by their position in the DOM (the natural tab order). Messing with that order does not help, but rather hinders  navigation.

The two legit uses for `tabindex`:
 
- Use `tabindex=0` to add an element to the natural tab order. There are primarily two reasons you would need to do this: 1) You are using an element that can't receive keyboard focus by default, such as using a `div` with `role="button"` instead of using a `button` element and 2) there are rare cases where you need to make something focusable that normally wouldn't need to be, for example, enabling keyboard users to scroll overflowing containers because Chrome and Safari still don't allow this by default.
- Add `tabindex="-1" to an element to allow it to receive programmatic focus (this does not place it in the natural tab order). This is used in a variety of situations for "focus management", such as when you need to set the focus on a heading, error message, or dialog, or you are building custom components such as tabbing interfaces or tree components.

The main thing I would stress is that you should never use a `tabindex` value greater than `0` (which leaves just `0` and `-1`) and that the natural tab order should be determined by the elements position in the DOM.

- **`accesskey` attribute**: used to define a keyboard shortcut for an element. This can help users with mobility impairments navigate the website more easily.

## WAI-ARIA, Roles, and Attributes

- **WAI-ARIA**: a set of attributes that can be added to HTML elements to improve accessibility. It provides additional information to assistive technologies about the purpose and structure of the content.

> - **Aria roles**: a set of predefined roles that can be added to HTML elements to define their purpose. This helps screen readers and other assistive technologies understand the content of the website. Examples include `role="button"`, `role="dialog"`, and `role="navigation"`.

Reviewer: Because I can't help it :smile: In a perfect world, there would never be the need to use `role="button"` because you would always use the `button` element instead. CSS is good enough nowadays that the arguments over being difficult to style a button are gone. The role of `dialog` is also going the same way now that the `dialog` element has full support in all browsers and is accessible out of the box. And of course the `nav` element has been around for a long time.

 If you want to humor me, you could use `role="tab", role="menu", and role="alert"` instead. These are all roles that don't have HTML equivalents. But I'm not insisting on it.

Author: I am looking at the associated lecture script and it looks like quite a few roles were mentioned. So we can update this review page to the three you mentioned.

> - **`aria-label` and `aria-labelledby` attributes**: The `aria-label` attribute can be used to provide a text alternative for an element. The `aria-labelledby` attribute can be used to associate an element with another element that serves as its label. This helps screen readers and other assistive technologies understand the purpose of the element.

Reviewer: ```suggestion
- **`aria-label` and `aria-labelledby` attributes**: These attributes are used to give an element a programmatic (or accessible name) which helps people using assistive technology (such as screen readers) understand the purpose of the element. They are often used when the visual label for an element is an image or symbol rather than text. `aria-label` allows you to define the name directly in the attribute while `aria-labelledby` allows you to reference existing text on the page.
```

Some some suggestions on how to make it a little clearer that these two attributes do the exact same thing, just in a slightly different way.

> - **`aria-hidden` attribute**: The `aria-hidden` attribute can be used to hide an element from screen readers. This can be useful for decorative elements that do not provide any meaningful content.

Reviewer: ```suggestion
- **`aria-hidden` attribute**: Used to hide an element from assistive technology such as screen readers. For example, this can be used to hide decorative images that do not provide any meaningful content.
```

Pure semantics here. While this is primarily used to hide content from screen reader users, it will also hide it from other types of assistive technology, basically anything that uses the accessibility tree API to communicate the contents on the page. If you think this distinction is too much or unnecessary, then I suggest it be changed to the following:

"Used to hide an element from screen reader users. For example, this can be used to hide decorative images that do not provide any meaningful content."

> - **`aria-expanded` attribute**: used to indicate whether an element is expanded or collapsed. This helps screen readers and other assistive technologies understand the state of the element.

Reviewer: ```suggestion
- **`aria-expanded` attribute**: Used to convey the state of a toggle (or disclosure) feature to screen reader users.
```

This attribute is always confusing because it goes on the control element but it is describing the state of the element it is controlling. So it doesn't describe the state of the element it is on, but rather the state of the element it is controlling. Seems clearer to just refer to both of these elements at the same time as a "toggle".

> - **`aria-live` attribute**: used to indicate that an element's content is live and should be announced by screen readers. This is useful for elements that are updated dynamically.

Reviewer: ```suggestion
- **`aria-live` attribute**: Used to indicate that an element's content is important enough to require that any changes to the content should be announced immediately to screen reader users. This can include status messages like updating the number of results returned from a search, or an error message displayed after an unsuccessful form submission.
```

Just being picky here. I didn't like that `aria-live` was being defined with the word "live".
